mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -1,121 +1,94 @@
-            })
-            .map(|(peer, state)| (*peer, state.clone()))
-            .collect();
-
-        let picked_peers_count = picked_peers.len();
-        if picked_peers_count < self.min_broadcast_recipient_count {
-            // randomly select fallback peers
-            // TODO add peer scoring schema to use for selecting fallback peers
-            let fallback_peers: PeerInfo = peer_info
-                .iter()
-                .filter(|(peer, state)| !picked_peers.contains_key(&peer) && state.is_alive)
-                .map(|(peer, state)| (*peer, state.clone()))
-                .choose_multiple(
-                    &mut rand::thread_rng(),
-                    self.min_broadcast_recipient_count - picked_peers_count,
-                )
-                .into_iter()
-                .collect();
+    pub fn update_peer_broadcast(&mut self, update: (PeerNetworkId, u64)) {
+        let (peer, new_timeline_id) = update;
+        self.peer_info.entry(peer).and_modify(|t| {
+            t.timeline_id = new_timeline_id;
+        });
+    }
 
-            picked_peers.extend(fallback_peers);
-        }
-        picked_peers
+    pub fn is_upstream_peer(&self, peer: PeerNetworkId) -> bool {
+        self.upstream_config.is_upstream_peer(peer)
     }
 
-    pub fn update_peer_broadcast(&self, peer_broadcasts: Vec<(PeerId, u64)>) {
-        let mut peer_info = self
-            .peer_info
-            .lock()
-            .expect("failed to acquire peer_info lock");
+    fn is_primary_upstream_peer(&self, peer: PeerNetworkId) -> bool {
+        self.upstream_config.is_primary_upstream_peer(peer)
+    }
 
-        for (peer_id, new_timeline_id) in peer_broadcasts {
-            peer_info.entry(peer_id).and_modify(|t| {
-                t.timeline_id = new_timeline_id;
-            });
+    // returns a peer's sync state if the peer is a chosen broadcast peer
+    pub fn get_picked_state(&self, peer: PeerNetworkId) -> Option<PeerSyncState> {
+        if self.is_picked_peer(peer) {
+            self.peer_info.get(&peer).cloned()
+        } else {
+            None
         }
     }
 
-    pub fn is_upstream_peer(&self, network_id: PeerId, peer_id: PeerId) -> bool {
-        self.upstream_config.is_upstream_peer(network_id, peer_id)
+    // checks whether a peer is a chosen broadcast recipient:
+    // - all primary peers
+    // - select fallback peers, if k-policy is enabled
+    // this does NOT check for whether this peer is alive
+    fn is_picked_peer(&self, peer: PeerNetworkId) -> bool {
+        if self.is_primary_upstream_peer(peer) {
+            return true;
+        }
+
+        let num_alive_primary_peers = self
+            .peer_info
+            .iter()
+            .filter(|(peer, state)| self.is_primary_upstream_peer(**peer) && state.is_alive)
+            .count();
+
+        // for fallback peers, broadcast if k-policy is on
+        num_alive_primary_peers < self.min_broadcast_recipient_count
```

Comment:
ok, so with this change we always broadcast to *all* fallback peers if num alive primaries < min_broadcast_recipient_count

btw, we probably shouldn't do that if all primaries are alive but their number is still less than min_broadcast_recipient_count. Imagine wallet use case where number of alive validators is 1, but  k for fallback is 3. You don't want to start broadcasting to secondaries right away